src/Interpreters/join_common.cpp

```diff
@@ -283,6 +285,87 @@ void addDefaultValues(IColumn & column, const DataTypePtr & type, size_t count)
         type->insertDefaultInto(column);
 }
 
+bool typesEqualUpToNullability(DataTypePtr left_type, DataTypePtr right_type)
+{
+    DataTypePtr left_type_strict = removeNullable(recursiveRemoveLowCardinality(left_type));
+    DataTypePtr right_type_strict = removeNullable(recursiveRemoveLowCardinality(right_type));
+    return left_type_strict->equals(*right_type_strict);
+}
+
+JoinConvertActions columnsNeedConvert(const Block & left_block, const Names & left_keys,
+                                      const Block & right_block, const Names & right_keys,
+                                      bool has_using)
+{
+    assert(left_keys.size() == right_keys.size());
+
+    /// only JOIN USING supported
+    if (!has_using)
+        return {};
+
+    JoinConvertActions actions;
+
+    Block left_block_dst = left_block;
+    Block right_block_dst = right_block;
+
+    std::unordered_set<std::string_view> visited_left;
+    std::unordered_set<std::string_view> visited_right;
+    bool any_need_cast = false;
+    for (size_t i = 0; i < left_keys.size(); ++i)
+    {
+        if (visited_left.contains(left_keys[i]) || visited_right.contains(right_keys[i]))
+        {
+            /// if one column joined with multiple different others do not perform conversion
+            /// e.g. `JOIN ... ON t1.a == t2.a AND t1.a == t2.b`
```

Comment:
We only support `USING`, is this possible?

Reply:
I've removed this code